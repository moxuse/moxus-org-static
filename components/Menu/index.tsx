/**
 * menu.jsx
 */
import * as React from "react";
import { useState, useRef, useEffect  } from 'react';
import Link from "next/link";
import { TweenMax } from 'gsap';

import { MenuContext } from "../../context/menu";

import styles from './styles.module.css';

const Menu: React.FC = (props) => {
  const { state, setState } = React.useContext(MenuContext);
  let menu = useRef(null);
  const [visibiiity, setVisibility] = useState(false);
  const [openAnimation, setOpenAnimation] = useState(null);
  const [closeAnimation, closeFadeAnimation] = useState(null);

  const onClick = (): void => {
    setState('close');
  }

  useEffect (() => {
    setOpenAnimation(
      TweenMax.to(menu, 0.6, {
          opacity: 1,
          ease: 'Power2.easeOut'
        }).pause()
    )
  },[]);

  useEffect (() => {
    closeFadeAnimation(
      TweenMax.to(menu, 0.6, {
          opacity: 0,
          ease: 'Power2.easeOut',
          onComplete: () => {
            setVisibility(false);
          }
        }).pause()
    )
  },[]);

  useEffect(() => {
    let value = 0;
    let onCompletion = undefined;
    if(state === 'open') {
      setVisibility(true);
      value = 0.8
    } else {
      onCompletion = () => { setVisibility(false); }
      value = 0;
    }  
    TweenMax.to(menu, 0.6, {
      opacity: value,
      ease: 'Power2.easeOut',
      onComplete: onCompletion
    }).play()
  })

  const visibleStyle = (state === 'close' && !visibiiity) ? styles.hide : '';

  return (
    <div className={`menu ${styles.menu} ${visibleStyle}`} ref={element => { menu = element }}>
      <ul onClick={onClick}>
        <li><Link href={`/`}>top</Link></li>
        <li><a href={`https://github.com/moxuse/cv/blob/master/README.md`} target='blank'>cv</a></li>
        <li><Link href={`/work`} >work</Link></li>
        <li><Link href={`/blog`} id={1}>blog</Link></li>
      </ul>
    </div>
  );
}

export default Menu;

// Menu.defaultProps = {
//   onClose: undefined
// };
