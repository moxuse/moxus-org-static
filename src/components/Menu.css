/**
 * menu.css
 */

.menu {
  z-index: 1;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  margin: 0 auto;
  background-color: #ffffff;
  opacity: 0.85;
  display: block;
}

.menu.hide {
  display: none;
}

.menu ul {
  margin: 0 auto;
  padding-top: 140px;
  padding-left: 0px;
  width: 141px;
  text-align: center;
}

.menu li {
  font-size: 38px;
  font-weight: bold;
  display: inline-block;
  padding: 30px 10px 30px 10px;
}

.menu a {
  text-decoration: none;
  color: #333;
}

.menu h1 {
  margin-top: 40px;
}


@media screen and (max-width: 768px) {
}

@media screen and (max-width: 480px) {
  .menu ul {
    padding-top: 60px;
  }
}
