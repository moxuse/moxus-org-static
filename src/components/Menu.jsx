/**
 * menu.jsx
 */
import React, { Component } from 'react';
import { Link } from "components/Router";
import { TweenLite } from 'gsap';

import styles from './Menu.css';

class Menu extends Component {
  constructor(props) {
    super(props);
    this.onClick = this.onClick.bind(this);
    this.setChangeState = this.setChangeState.bind(this);
  }

  onClick() {
    this.props.onClose();
    document.getElementById('.menu_btn').classList.remove('open');
  }

  openAnimation() {
    this.refs.menu.classList.remove('hide');
    TweenLite.set(this.refs.menu, {opacity: 0});
    TweenLite.to(this.refs.menu, 0.6, {
      opacity: 1,
      ease: 'Power2.easeOut'
    });
    Array.from(this.refs.menu_list.querySelectorAll('li')).forEach((item, i) => {
      TweenLite.set(item, {y: 80, opacity: 0});
      TweenLite.to(item, 0.45, {
        y: 1,
        opacity: 1,
        delay: (i * 0.03),
        ease: 'Power2.easeInOut'
      });
    });
  }

  cloaseAnimation() {
    TweenLite.to(this.refs.menu, 0.6, {
      opacity: 0,
      delay: 0.2,
      ease: 'Power2.easeOut',
      onComplete: () => this.refs.menu.classList.add(styles.hide)
    });

    Array.from(this.refs.menu_list.querySelectorAll('li')).forEach((item, i) => {
      TweenLite.set(item, {y: 0});
      TweenLite.to(item, 0.45, {
        y: -80,
        opacity: 0,
        delay: (i * 0.03),
        ease: 'Power2.easeInOut'
      });
    });
  }

  setChangeState(state) {
    switch (state) {
      case 'open':
        this.refs.menu.classList.remove(styles.hide);
        this.openAnimation();
        break;

      case 'close':
        this.cloaseAnimation();
        break;
      default:
        break;
    }
  }

  render() {
    return (
      <div className={`menu ${styles.menu} ${styles.hide}`} ref='menu'>
        <ul ref='menu_list'>
          <li><Link to={`/`} onClick={this.onClick}>top</Link></li>
          <li><a href={`https://github.com/moxuse/cv/blob/master/README.md`} target='blank'>cv</a></li>
          <li><Link to={`/work`} onClick={this.onClick}>work</Link></li>
          <li><Link to={`/blog`} id={1} onClick={this.onClick}>blog</Link></li>
        </ul>
      </div>
    );
  }
}

export default Menu;

// Menu.defaultProps = {
//   onClose: undefined
// };
