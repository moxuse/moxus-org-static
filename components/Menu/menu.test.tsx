import '@testing-library/jest-dom'
import * as React from 'react'
import {render, fireEvent, screen} from '@testing-library/react'
import Menu from '.';

describe('Menu', () => {
  it('render correctly', async () => {
    render(<Menu />);

    expect(await screen.getByText('work')).toBeVisible();

        // screen.getAllByRole('listitem');

    // screen.debug();
  });
})
