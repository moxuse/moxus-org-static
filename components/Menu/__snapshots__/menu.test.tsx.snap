// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly 1`] = `
<div
  className="menu undefined undefined"
>
  <ul
    onClick={[Function]}
  >
    <li>
      <a
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        top
      </a>
    </li>
    <li>
      <a
        href="https://github.com/moxuse/cv/blob/master/README.md"
        target="blank"
      >
        cv
      </a>
    </li>
    <li>
      <a
        href="/work"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        work
      </a>
    </li>
    <li>
      <a
        href="/blog"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        blog
      </a>
    </li>
  </ul>
</div>
`;
